File: example_code/topic2/metropolis.py
# estimate Gaussian integral using the Monte Carlo algorithm

from math import exp, pi, sqrt
from random import random

def P(x):
    return exp(- x**2 / 2) / sqrt(2 * pi)

def f_over_w(x):
    return x**2

x = 0.0                         # initial position of walker
delta = 1.0                     # step size

accepts = 0                     # number of steps accepted

def MetropolisStep():
    global x, accepts           # updated the global value
    xTrial = x + (2 * random() - 1) * delta
    w = P(xTrial) / P(x)
    if w >= 1:                  # uphill
        x = xTrial              # accept the step
        accepts += 1
    else:                       # downhill
        if random() < w:        # but not too far
            x = xTrial          # accept the step
            accepts += 1

# get input parameters from user
delta = float(input(" Enter step size  delta: "))
M = int(input(" Enter number of trials: "))
N = int(input(" Enter steps per trial: "))

fSum = 0.0                      # accumulator for <f>
sqdSum = 0.0                    # accumulator for <f> * <f>
errSum = 0.0                    # accumulator error estimates
for i in range(M):
    avg = 0.0                   # accumulator for f(x)
    var = 0.0                   # accumulator for f(x) * f(x)
    for j in range(N):
        MetropolisStep()
        fx = f_over_w(x)
        avg += fx
        var += fx**2
    avg /= N
    var /= N
    var = var - avg**2
    err = sqrt(var / N)
    fSum += avg
    sqdSum += avg**2
    errSum += err
ans = fSum / M
stdDev = sqrt(sqdSum / M - ans**2)
stdDev /= sqrt(float(M))
err = errSum / M
err /= sqrt(float(M))
print("\n Exact  answer: " + str(1.0))
print("      Integral: " + str(ans) + "  +-  " + str(err))
print("     Std. Dev.: " + str(stdDev))
print("  Accept ratio: " + str(accepts / float(N * M)))

# under construction

File: example_code/topic2/metropolis.cpp
// estimate Gaussian integral using the Monte Carlo algorithm

#include <cmath>
#include <cstdlib>
#include <iostream>
using namespace std;

const double pi = 4 * atan(1.0);

inline double std_rand()
{
    return rand() / (RAND_MAX + 1.0);
}

double P(double x) {
    return exp(- x * x / 2) / sqrt(2 * pi);
}

double f_over_w(double x) {
    return x * x;
}

double x = 0;             // initial position of walker
double delta = 1.0;       // step size

int accepts = 0;          // number of steps accepted

void MetropolisStep() {
    double xTrial = x + (2 * std_rand() - 1) * delta;
    double w = P(xTrial) / P(x);
    if (w >= 1) {                  // uphill
        x = xTrial;                // accept the step
        ++accepts;
    } else {                       // downhill
        if (std_rand() < w) {      // but not too far
            x = xTrial;            // accept the step
            ++accepts;
        }
    }
}

int main() {

    // get input parameters from user
    cout << "Enter step size  delta: ";
    cin >> delta;
    cout << "Enter number of trials: ";
    int M;
    cin >> M;
    cout << "Enter steps per  trial: ";
    int N;
    cin >> N;

    double sum = 0;        // accumulator for <f>
    double sqdSum = 0;     // accumulator for <f> * <f>
    double errSum = 0;     // accumulator error estimates
    for (int i = 0; i < M; i++) {
        double avg = 0;        // accumulator for f(x)
        double var = 0;        // accumulator for f(x) * f(x)
        for (int j = 0; j < N; j++) {
            MetropolisStep();
            double fx = f_over_w(x);
            avg += fx;
            var += fx * fx;
        }
        avg /= N;
        var /= N;
        var = var - avg * avg;
        double err = sqrt(var / N);
        sum += avg;
        sqdSum += avg * avg;
        errSum += err;
    }
    double ans = sum / M;
    double stdDev = sqrt(sqdSum / M - ans * ans);
    stdDev /= sqrt(double(M));
    double err = errSum / M;
    err /= sqrt(double(M));
    cout << "\n Exact  answer: " << 1.0 << endl;
    cout << "      Integral: " << ans << "  +-  " << err << endl;
    cout << "     Std. Dev.: " << stdDev << endl;
    cout << "  Accept ratio: " << accepts / double(N * M) << endl;
}
